biogeochem/EDPhysiologyMod.F90

```diff
@@ -503,42 +553,111 @@ subroutine trim_canopy( currentSite )
-                      endif
+                ! Construct the arrays for a least square fit of the net_net_uptake versus the cumulative lai
+                if (currentCohort%nv > nll .and. currentCohort%nv - z < nll) then  ! Only for nll layers
+                  nnu_clai_a(1,1) = nnu_clai_a(1,1) + 1 ! Increment for each layer used
+                  nnu_clai_a(1,2) = nnu_clai_a(1,2) + currentCohort%year_net_uptake(z) - currentCohort%leaf_cost
+                  nnu_clai_a(2,1) = nnu_clai_a(1,2)
+                  nnu_clai_a(2,2) = nnu_clai_a(2,2) + (currentCohort%year_net_uptake(z) - currentCohort%leaf_cost)**2
+                  nnu_clai_b(1,1) = nnu_clai_b(1,1) + cumulative_lai
+                  nnu_clai_b(2,1) = nnu_clai_b(2,1) + (cumulative_lai * & 
+                                    (currentCohort%year_net_uptake(z) - currentCohort%leaf_cost))
+                end if
+
+                ! Check leaf cost against the yearly net uptake for that cohort leaf layer
+                if (currentCohort%year_net_uptake(z) < currentCohort%leaf_cost) then
+                   ! Make sure the cohort trim fraction is great than the pft trim limit
+                   if (currentCohort%canopy_trim > EDPftvarcon_inst%trim_limit(ipft)) then
+
+                     !  if ( debug ) then
+                     !     write(fates_log(),*) 'trimming leaves', &
+                     !           currentCohort%canopy_trim,currentCohort%leaf_cost
+                     !  endif
 
                       ! keep trimming until none of the canopy is in negative carbon balance.              
-                      if (currentCohort%hite > EDPftvarcon_inst%hgt_min(ipft))then
+                      if (currentCohort%hite > EDPftvarcon_inst%hgt_min(ipft)) then
                          currentCohort%canopy_trim = currentCohort%canopy_trim - &
                                EDPftvarcon_inst%trim_inc(ipft)
-                         if (EDPftvarcon_inst%evergreen(ipft) /= 1)then
+
+                         if (EDPftvarcon_inst%evergreen(ipft) /= 1) then
                             currentCohort%laimemory = currentCohort%laimemory * &
                                   (1.0_r8 - EDPftvarcon_inst%trim_inc(ipft)) 
                          endif
+
                          trimmed = .true.
-                      endif
-                   endif
-                endif
-             endif !leaf activity? 
-          enddo !z
 
+                      endif ! hite check
+                   endif ! trim limit check
+                endif ! net uptake check
+             endif ! leaf activity check 
+          enddo ! z, leaf layer loop
+
+          ! Compute the optimal cumulative lai based on the cohort net-net uptake profile if at least 2 leaf layers
+          if (nnu_clai_a(1,1) > 1) then
+
+            ! Compute the optimum size of the work array
+            lwork = -1 ! Ask sgels to compute optimal number of entries for work
```

Comment:
checked through the lapack dgels description.  Verified that nnu_cla_b is not overwritten when work dne 0

Reply:
could we add some documentation here describing the arguments that dgels is expecting, etc, like what we have for the lapack dgesv call here: https://github.com/NGEET/fates/blob/master/biogeophys/FatesPlantHydraulicsMod.F90#L4812?

Reply:
actually i see now that's what is going on in the code starting at line 400 above... sorry.

Reply:
At @rgknox suggestion I am writing up a technote to help describe the setup: https://github.com/glemieux/fates-jupyter/blob/develop/leaf-flutter/pr623-linear-solver.ipynb

Reply:
I also like the idea of some comments in line.  Something that would explain what variables are equated to the matrix terms: ie  Y = X * b, and how the X and Y terms are manipulated into arguments of dgels()

Reply:
Roger that.  I'll flesh out the comments in that part.